File: 286-walls-and-gates/286-walls-and-gates.java
class Solution {
    public void wallsAndGates(int[][] rooms) {
        int INF = Integer.MAX_VALUE;
        int GATE = 0;
        int WALL = -1;
        Queue<int[]> q = new LinkedList<>();
        
        for (int row=0; row < rooms.length; row++) {
            for (int col=0; col < rooms[0].length; col++) {
                if (rooms[row][col] == GATE)
                    q.add(new int[] {row, col, 0} );
            };
        };
        
        int[] dcol = {1, -1, 0, 0};
        int[] drow = {0, 0, 1, -1};
        while (!q.isEmpty()) {
            int[] curr = q.remove();
            int row = curr[0];
            int col = curr[1];
            int dist = curr[2];
            
            for (int count = 0; count < 4; count++) {
                int new_row = row + drow[count];
                int new_col = col + dcol[count];
                
                if (new_row < 0 || new_col < 0 || new_row >= rooms.length || new_col >= rooms[0].length) {
                    continue;
                } else if (rooms[new_row][new_col] == INF) {
                    rooms[new_row][new_col] = dist+1;
                    q.add(new int[] {new_row, new_col, dist+1});
                };
                
                
            };
            
        };
        
        
    };
    
    
}

File: 286-walls-and-gates/286-walls-and-gates.py
class Solution:
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        """
        INF = 2**31 - 1
        GATE = 0
        WALL = -1
        q = collections.deque()
        directs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def bfs(q):
            while q:
                row, col, dist = q.popleft()
                for add_row, add_col in directs:
                    new_row = row + add_row
                    new_col = col + add_col
                    if new_row < 0 or new_col < 0 or new_row >= len(rooms) or new_col >= len(rooms[0]):
                        continue
                    elif rooms[new_row][new_col] == INF:
           
                        rooms[new_row][new_col] = min(rooms[new_row][new_col], dist+1)
                        q.append([new_row, new_col, dist+1])
                        
                
            
        for row in range(len(rooms)):
            for col in range(len(rooms[0])):
                if rooms[row][col] == GATE:
                    q.append([row, col, 0])
                    
        if q:
            bfs(q)

        return rooms